src/main/java/org/aksw/iguana/cc/worker/impl/SPARQLProtocolWorker.java

```diff
@@ -0,0 +1,331 @@
+                                executionStats.add(execution);
+                        }
+
+                    }
+                } else if (config().completionTarget() instanceof TimeLimit timeLimit) {
+                    final Instant endTime = Instant.now().plus(timeLimit.duration());
+                    Instant now;
+                    while ((now = Instant.now()).isBefore(endTime)) {
+                        final Duration timeToEnd = Duration.between(now, endTime);
+                        final boolean timeoutBeforeEnd = config().timeout().compareTo(timeToEnd) < 0;
+                        final Duration thisQueryTimeOut = (timeoutBeforeEnd) ? config().timeout() : timeToEnd;
+                        // If timeoutBeforeEnd is false, fail shouldn't be counted as timeout
+                        ExecutionStats execution = executeQuery(thisQueryTimeOut, !timeoutBeforeEnd);
+                        if (execution != null)
+                            executionStats.add(execution);
+                    }
+                }
+            } catch (IOException | URISyntaxException e) {
+                throw new RuntimeException(e);
+            }
+
+            return new Result(this.workerId, executionStats);
+        }, executor);
+    }
+
+    private ExecutionStats executeQuery(Duration timeout, boolean discardOnFailure) throws IOException, URISyntaxException {
+        HttpExecutionResult result = executeHttpRequest(timeout);
+        Optional<Integer> statuscode = Optional.empty();
+        if (result.response().isPresent())
+            statuscode = Optional.of(result.response().get().statusCode());
+
+        if (result.successful()) { // 2xx
+            // process result
+            if (!responseBodyProcessor.add(result.actualContentLength().getAsLong(), result.hash().getAsLong(), result.outputStream().get())) {
+                this.responseBodybbaos = result.outputStream().get();
+            } else {
+                this.responseBodybbaos = new BigByteArrayOutputStream();
+            }
+        }
+
+        if (!result.completed() && discardOnFailure) {
+            return null;
+        }
+
+        return new ExecutionStats(
+                result.queryID(),
+                result.requestStart(),
+                result.duration(),
+                statuscode,
+                result.actualContentLength(),
+                result.hash,
+                result.exception()
+        );
+    }
+
+
+    private HttpExecutionResult executeHttpRequest(Duration timeout) throws IOException, URISyntaxException {
+        final QueryHandler.QueryStreamWrapper queryHandle = config().queries().getNextQueryStream();
+        final HttpRequest request = requestFactory.buildHttpRequest(
+                queryHandle.queryInputStream(),
+                timeout,
+                config().connection(),
+                config().acceptHeader()
+        );
+
+        final Instant requestStart = Instant.now();
```

Comment:
Before we take the requestStart instant, we must make sure that there is a thread in the pool. This should be done with [`pool.getActiveCount() < 2`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#getActiveCount--).

Reply:
Could you elaborate, which funtion you're mentioning?

Reply:
see below